php_phongo.c

```diff
@@ -40,6 +40,10 @@
 #include <ext/standard/php_smart_str.h>
 #endif
 
+#ifdef MONGOC_ENABLE_CLIENT_SIDE_ENCRYPTION
+#include <mongocrypt/mongocrypt.h>
```

Comment:
I reverted the changes of [MONGOCRYPT-213](https://jira.mongodb.org/browse/MONGOCRYPT-213) because in moving the public headers, I modified their relative #include's, which inadvertently broke the Go driver build (which is not build in the libmongocrypt evergreen project). So I'm considering other solutions to avoid any disruption.

But I'm not sure I fully understand the problem here. Is the purpose of `libmongocrypt-tmp` directory to make it possible to refer to the public header as `#include <mongocrypt/mongocrypt.h>`? Would this work without any changes to libmongocrypt if this  were `#include <mongocrypt.h>`, or is that undesirable?

If it is necessary, perhaps it would suffice if libmongocrypt have a single forwarding header in a mongocrypt subdirectory. I.e. I simply add `src/mongocrypt/mongocrypt.h`, which just includes `../mongocrypt.h`.

Reply:
> But I'm not sure I fully understand the problem here. Is the purpose of `libmongocrypt-tmp` directory to make it possible to refer to the public header as `#include <mongocrypt/mongocrypt.h>`? Would this work without any changes to libmongocrypt if this were `#include <mongocrypt.h>`, or is that undesirable?

Changing this to `#include <mongocrypt.h>` would work for our use-case, as that would allow us to continue specifying `src/libmongocrypt/src` as include path.

A forwarding header could also work. Since we already need to ship a `mongocrypt-export.h` (PHP extensions don't use cmake), we can also add that file to avoid needing to package it. I can test that solution, maybe we can avoid making upstream changes for this, unless you want to solve this for all projects that bundle libmongocrypt.

Reply:
> Changing this to #include <mongocrypt.h> would work for our use-case, as that would allow us to continue specifying src/libmongocrypt/src as include path.

If forwarding headers were added, would `src/libmongoc/src` no longer be in the include path? I had thought `#include <mongocrypt/mongocrypt.h>` intends to resolve to src/libmongoc/src/mongocrypt/mongocrypt.h, so `src/libmongoc/src` needs to be in the include path regardless.

Also, I now remember we made a similar change to libmongoc/libbson. Prior to https://jira.mongodb.org/browse/CDRIVER-2800, consumers were expected to include like `#include <bson.h>`, but that change added the `bson`/`mongoc` prefix directories, and changed the top-level files to be forwarding headers.

No other drivers bundle libmongocrypt currently. It's certainly possible to change libmongocrypt to be included via `mongocrypt/mongocrypt.h` in both the source and the installed headers, but I'm just not sure if the advantage is just convention or a real necessity (e.g. how CDRIVER-2800 was necessary for Darwin frameworks).

An easy and risk-free change in libmongocrypt would be to add forwarding headers. Let me know if you'd like me to make that change in libmongocrypt, or if you'd rather do that here.

Reply:
I've worked around this for now by adding our own mongocrypt.h file: https://github.com/mongodb/mongo-php-driver/pull/1065/files#diff-f309cffeaf85fbe6fe0f72f5637ffe50

For now this works, but not needing compat files here would sure be better.

Another issue is mongocrypt-export.h, which normally is generated using cmake. Since PHP extensions use make, I've copied a mongocrypt-export.h into our compat folder, but ideally we wouldn't have to have this ourselves either. However, I can live with keeping it here for the time being.

Reply:
I'll defer to @kevinAlbs - not sure what we can do about mongocrypt-export.h, but a forwarding file for mongocrypt.h would certainly be nice.

Reply:
To check, the change in libmongocrypt would be to effectively have in the src directory  what you're doing with src/libmongocrypt-compat? I.e. add the directory mongocrypt and these forwarding headers?
1. src/mongocrypt/mongocrypt.h
2. src/mongocrypt/mongocrypt-export.h

As @alcaeus mentioned, mongocrypt-export.h is generated by cmake (and I believe is system dependent). Including the mongocrypt-export.h included here, with empty `MONGOCRYPT_EXPORT`/`MONGOCRYPT_NO_EXPORT` as part of libmongocrypt's src seems a little odd, but also harmless.

Reply:
Yes, those headers would work fine for us. If cmake overwrites the default mongocrypt/mongocrypt-export.h, having it there with empty macros would indeed be harmless 👍

Reply:
Adding the pregenerated mongocrypt-export.h and forwarding mongocrypt.h header to libmongocrypt still seems a little odd to me. It's only necessary for this bundling, and we wouldn't be testing them within libmongocrypt. Though we can always revisit this later, I vote we leave these in the mongo-php-driver repo for now.